src/che-theia-github-plugin-backend-plugin.ts

```diff
@@ -0,0 +1,103 @@
+/*********************************************************************
+ * Copyright (c) 2018 Red Hat, Inc.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ **********************************************************************/
+
+import * as theia from '@theia/plugin';
+import { GithubServiceImpl } from './node/github-service-impl';
+import { WsMasterHttpClient } from './node/ws-master-http-client';
+import { Credentials, SshKeyPair } from './common/github-model';
+
+const githubService = new GithubServiceImpl();
+
+let credentialsObject: Credentials;
+
+export function start(context: theia.PluginContext) {
+    const GENERATE_AND_UPLOAD: theia.Command = {
+        id: 'github:generate_and_upload',
+        label: 'GitHub: Generate Ssh key and upload it to GitHub ...'
+    };
+    const AUTHENTICATE: theia.Command = {
+        id: 'github:authenticate',
+        label: 'GitHub: Authenticate ...'
+    };
+
+    context.subscriptions.push(
+        theia.commands.registerCommand(GENERATE_AND_UPLOAD, async () => {
+            if (!credentialsObject) {
+                await authenticate();
+            }
+            uploadSshKey(credentialsObject);
+        })
+    );
+    context.subscriptions.push(
+        theia.commands.registerCommand(AUTHENTICATE, () => {
+            authenticate();
+        })
+    );
+}
+
+async function uploadSshKey(credentials: Credentials): Promise<void> {
+    const title = 'theia';
+    const publicKey = await getOrGenerateSshKey();
+    await githubService.deleteSshKey(credentials, title);
+    try {
+        if (publicKey) {
+            await githubService.uploadSshKey(credentials, title, publicKey);
+            theia.window.showInformationMessage('Public SSH key was successfully uploaded to GitHub');
+        }
+    } catch (error) {
+        theia.window.showErrorMessage(JSON.parse(error.message).message);
+    }
+}
+
+async function getOrGenerateSshKey(): Promise<string | undefined> {
+    const cheApi = await theia.env.getEnvVariable('CHE_API_INTERNAL');
+    const machineToken = await theia.env.getEnvVariable('CHE_MACHINE_TOKEN');
+    try {
+        const request = await new WsMasterHttpClient(cheApi).get<SshKeyPair>(`/ssh/vcs/find?name=github.com` + (machineToken ? '&token=' + machineToken : ''));
```

Comment:
not sure this is the best way to pass the token
https://github.com/eclipse/che-theia-terminal-extension/blob/master/che-theia-terminal/src/node/workspace-service-impl.ts#L129

Reply:
done